default/python/universe_generation/specials.py

```diff
@@ -1,77 +1,178 @@
 import random
+from collections import defaultdict
+
 import freeorion as fo
 import statistics
 import universe_tables
+from timers import Timer
+from galaxy_topology import get_systems_within_jumps
+
+REPEAT_RATE = {1: 0.08, 2: 0.05, 3: 0.01, 4: 0.00}
+
+def calc_num_placed(objs):
+    """Return a list of number placed at each obj"""
+    return [1 if random.random() > REPEAT_RATE[1] else
+            2 if random.random() > REPEAT_RATE[2] else
+            3 if random.random() > REPEAT_RATE[3] else 1
+            for __ in objs]
+
+
+def place_special_fast(specials, obj, specials_timer, per_special_timer):
+    """ Place at least single special.
+    Return the number of specials placed.
+    """
+    num_placed = 0
+    # Calculate the conditional probabilities that each special is
+    # placed here given that a special will be placed here.
+    probs = [fo.special_spawn_rate(sp) for sp in specials]
+    total_prob = sum(probs)
+    thresholds = [x/total_prob for x in probs]
+
+    chance = random.random()
+    for threshold, special in zip(thresholds, specials):
+        if chance > threshold:
+            chance -= threshold
+            continue
+
+        per_special_timer.start(special + "_place")
+        specials_timer.start("Place Special")
+        # All prerequisites and the test have been met, now add this special to this universe object.
+        fo.add_special(obj, special)
+        num_placed += 1
+        print "Special", special, "added to", fo.get_name(obj)
 
+        break
 
+    per_special_timer.start("None")
+    return num_placed
```

Comment:
I am not sure what you are trying to achieve here.

It seems to me that this function will always place exactly one special - the probabilities are normalized to 1 so chance wil eventually be greater than some threshold. After placing one special, you break out of the loop.

So num_placed will always be 1, making it both a pointless variable and return value. Docstring specifies the functio should be able to place more than 1 special. Am I missing something here?